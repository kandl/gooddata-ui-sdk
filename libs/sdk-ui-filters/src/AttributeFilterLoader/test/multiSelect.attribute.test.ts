// (C) 2019-2022 GoodData Corporation
import { newTestAttributeFilterHandler } from "./fixtures";
import { waitForAsync } from "./testUtils";

describe("MultiSelectAttributeFilterHandler", () => {
    it("should init attribute", async () => {
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        attributeFilterHandler.init("init");

        await waitForAsync();

        expect(attributeFilterHandler.getAttribute()).toMatchSnapshot();
    });

    it("should load attribute and trigger success callback", async () => {
        const onAttributeLoadStart = jest.fn();
        const onAttributeLoadSuccess = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        expect(attributeFilterHandler.getAttributeStatus()).toMatchSnapshot("getAttributeStatus before load");

        attributeFilterHandler.init();
        attributeFilterHandler.onAttributeLoadStart(onAttributeLoadStart);
        attributeFilterHandler.onAttributeLoadSuccess(onAttributeLoadSuccess);
        attributeFilterHandler.loadAttribute("attributeSuccess");

        expect(attributeFilterHandler.getAttributeStatus()).toMatchSnapshot("getAttributeStatus during load");

        await waitForAsync();

        expect(attributeFilterHandler.getAttributeStatus()).toMatchSnapshot("getAttributeStatus after load");

        expect(attributeFilterHandler.getAttribute()).toMatchSnapshot();
        expect(onAttributeLoadStart).toHaveBeenCalledTimes(1);
        expect(onAttributeLoadSuccess).toHaveBeenCalledTimes(1);
        expect(onAttributeLoadStart.mock.calls[0]).toMatchSnapshot("onAttributeLoadStart parameters");
        expect(onAttributeLoadSuccess.mock.calls[0]).toMatchSnapshot("onAttributeLoadSuccess parameters");
    });

    it("should load attribute and trigger error callback", async () => {
        const onAttributeLoadStart = jest.fn();
        const onAttributeLoadError = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("nonExisting");

        expect(attributeFilterHandler.getAttributeStatus()).toMatchSnapshot("getAttributeStatus before load");

        attributeFilterHandler.init();

        await waitForAsync();

        attributeFilterHandler.onAttributeLoadStart(onAttributeLoadStart);
        attributeFilterHandler.onAttributeLoadError(onAttributeLoadError);
        attributeFilterHandler.loadAttribute("attributeError");

        expect(attributeFilterHandler.getAttributeStatus()).toMatchSnapshot("getAttributeStatus during load");

        await waitForAsync();

        expect(attributeFilterHandler.getAttributeStatus()).toMatchSnapshot("getAttributeStatus after load");

        expect(attributeFilterHandler.getAttribute()).toMatchSnapshot();
        expect(onAttributeLoadStart).toHaveBeenCalledTimes(1);
        expect(onAttributeLoadError).toHaveBeenCalledTimes(1);
        expect(onAttributeLoadStart.mock.calls[0]).toMatchSnapshot("onAttributeLoadStart parameters");
        expect(onAttributeLoadError.mock.calls[0]).toMatchSnapshot("onAttributeLoadError parameters");
    });

    it("should load attribute and trigger cancel callback on another attribute call", async () => {
        const onAttributeLoadStart = jest.fn();
        const onAttributeLoadCancel = jest.fn();
        const attributeFilterHandler = newTestAttributeFilterHandler("positive");

        expect(attributeFilterHandler.getAttributeStatus()).toMatchSnapshot("getAttributeStatus before load");

        attributeFilterHandler.init();

        await waitForAsync();

        attributeFilterHandler.onAttributeLoadStart(onAttributeLoadStart);
        attributeFilterHandler.onAttributeLoadCancel(onAttributeLoadCancel);
        attributeFilterHandler.loadAttribute("attributeToBeCanceled");

        expect(attributeFilterHandler.getAttributeStatus()).toMatchSnapshot("getAttributeStatus during load");

        attributeFilterHandler.loadAttribute("attributeSuccess");

        expect(attributeFilterHandler.getAttributeStatus()).toMatchSnapshot(
            "getAttributeStatus during another load",
        );

        await waitForAsync();

        expect(attributeFilterHandler.getAttributeStatus()).toMatchSnapshot("getAttributeStatus after load");

        expect(attributeFilterHandler.getAttribute()).toMatchSnapshot();
        expect(onAttributeLoadStart).toHaveBeenCalledTimes(2);
        expect(onAttributeLoadCancel).toHaveBeenCalledTimes(1);
        expect(onAttributeLoadStart.mock.calls[0]).toMatchSnapshot("first onAttributeLoadStart parameters");
        expect(onAttributeLoadStart.mock.calls[1]).toMatchSnapshot("second onAttributeLoadStart parameters");
        expect(onAttributeLoadCancel.mock.calls[0]).toMatchSnapshot("onAttributeLoadCancel parameters");
    });
});
