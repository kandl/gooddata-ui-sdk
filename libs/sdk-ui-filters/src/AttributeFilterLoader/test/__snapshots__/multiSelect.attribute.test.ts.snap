// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelectAttributeFilterHandler should init attribute 1`] = `
Object {
  "description": "",
  "displayForms": undefined,
  "id": "attr.product.id",
  "production": true,
  "ref": Object {
    "uri": "/gdc/md/referenceworkspace/obj/1054",
  },
  "title": "Product",
  "type": "attribute",
  "uri": "/gdc/md/referenceworkspace/obj/1054",
}
`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger cancel callback on another attribute call 1`] = `
Object {
  "description": "",
  "displayForms": undefined,
  "id": "attr.product.id",
  "production": true,
  "ref": Object {
    "uri": "/gdc/md/referenceworkspace/obj/1054",
  },
  "title": "Product",
  "type": "attribute",
  "uri": "/gdc/md/referenceworkspace/obj/1054",
}
`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger cancel callback on another attribute call: first onAttributeLoadStart parameters 1`] = `
Array [
  Object {
    "correlation": "attributeToBeCanceled",
  },
]
`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger cancel callback on another attribute call: getAttributeStatus after load 1`] = `"success"`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger cancel callback on another attribute call: getAttributeStatus before load 1`] = `"pending"`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger cancel callback on another attribute call: getAttributeStatus during another load 1`] = `"loading"`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger cancel callback on another attribute call: getAttributeStatus during load 1`] = `"loading"`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger cancel callback on another attribute call: onAttributeLoadCancel parameters 1`] = `
Array [
  Object {
    "correlation": "attributeToBeCanceled",
  },
]
`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger cancel callback on another attribute call: second onAttributeLoadStart parameters 1`] = `
Array [
  Object {
    "correlation": "attributeSuccess",
  },
]
`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger error callback 1`] = `undefined`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger error callback: getAttributeStatus after load 1`] = `"error"`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger error callback: getAttributeStatus before load 1`] = `"pending"`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger error callback: getAttributeStatus during load 1`] = `"loading"`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger error callback: onAttributeLoadError parameters 1`] = `
Array [
  Object {
    "correlation": "attributeError",
    "error": [Error: No attribute recording non-existing-displayForm],
  },
]
`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger error callback: onAttributeLoadStart parameters 1`] = `
Array [
  Object {
    "correlation": "attributeError",
  },
]
`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger success callback 1`] = `
Object {
  "description": "",
  "displayForms": undefined,
  "id": "attr.product.id",
  "production": true,
  "ref": Object {
    "uri": "/gdc/md/referenceworkspace/obj/1054",
  },
  "title": "Product",
  "type": "attribute",
  "uri": "/gdc/md/referenceworkspace/obj/1054",
}
`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger success callback: getAttributeStatus after load 1`] = `"success"`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger success callback: getAttributeStatus before load 1`] = `"pending"`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger success callback: getAttributeStatus during load 1`] = `"loading"`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger success callback: onAttributeLoadStart parameters 1`] = `
Array [
  Object {
    "correlation": "attributeSuccess",
  },
]
`;

exports[`MultiSelectAttributeFilterHandler should load attribute and trigger success callback: onAttributeLoadSuccess parameters 1`] = `
Array [
  Object {
    "attribute": Object {
      "description": "",
      "displayForms": undefined,
      "id": "attr.product.id",
      "production": true,
      "ref": Object {
        "uri": "/gdc/md/referenceworkspace/obj/1054",
      },
      "title": "Product",
      "type": "attribute",
      "uri": "/gdc/md/referenceworkspace/obj/1054",
    },
    "correlation": "attributeSuccess",
  },
]
`;
